### DISCUSSION TITLE
Use of the ATI nano ft sensor with YARP in Linux

### DISCUSSION THREAD

**Discussion by 2103simon** [robotology/community#564] [Feedback & Meta]
Dear robotology community,

I want to work on a proper characterization of the current (and future) iCub fingertip(s) to tune my gazebo skin plugin. First results I acquired by using the [Omega.3](https://www.forcedimension.com/products/omega) robot from force dimensions suggest that the internal force measurement of the mentioned robot is not accurate enough. To overcome this issue I want to use the [ATI nano ft sensor](https://www.ati-ia.com/products/ft/ft_models.aspx?id=Nano17). Further the [NI 6210](https://www.ni.com/it-it/support/model.usb-6210.html) is used as AD converter. I found [this repo](https://github.com/robotology/forcetorque-yarp-devices) providing drivers for the sensor, but after trying a couple of days to get i t running I must face the truth: I can't manage it. I am not able to install the needed drivers on my Linux OS. 

Following my question and my request:
Knowing, that on iCub a similar sensor to the ATI nano is used (just bigger), I assume there must be an option to NOT use the NI 6210 to read the analog data, but use the same pipeline as on iCub. That would hold multiple benefits. 
1. For future use of the ft sensor (nano or not) a tested implementation in YARP is available (which allows easy closed loop experiments with force feedback/measurement). 
2. A testing procedure to analyze/characterize new fingertip(prototypes) and 
3. with that a way to tune skin plugins according to the physical sensor.

I am very happy about any suggestions and feedback!

Best
Simon

**Comment by traversaro** [comment]
> I found [this repo](https://github.com/robotology/forcetorque-yarp-devices) providing drivers for the sensor, but after trying a couple of days to get i t running I must face the truth: I can't manage it. I am not able to install the needed drivers on my Linux OS.

Just to clarify, the [`ati_ethernet`](https://github.com/robotology/forcetorque-yarp-devices/tree/v0.2.0/ATI_Ethernet) YARP device driver provided in that repo is written to communicate with a "Net F/T" electronics from ATI (see https://www.ati-ia.com/products/ft/ft_SystemInterfaces.aspx). If you do not have that electronics to connect to the ATI sensor, I am afraid it will be of little use.

**Comment by 2103simon** [comment]
Thanks for the fast reply! But doesn't the repo also contains the ft sensor [here](https://github.com/robotology/forcetorque-yarp-devices/tree/master/ftSensor). It is not fully clear to which kind of sensor the name 'forcetorque-yarp-devices/ftSensor' is referring. Nevertheless, the ft sensors used in iCub are from the same company (as mentioned, just bigger). So basically there has to be a way to read them out using YARP, or do you use another middle ware and bridge to YARP?

**Reply by traversaro** [reply]
> Nevertheless, the ft sensors used in iCub are from the same company (as mentioned, just bigger).

No, the FT sensors used in the iCub are custom produced by IIT, see https://icub-tech-iit.github.io/documentation/ft_sensors/ . The related YARP device drivers are [`canBusAnalogSensor`](https://github.com/robotology/icub-main/tree/v1.22.0/src/libraries/icubmod/canBusAnalogSensor) that read the FT directly via a CAN bus or [`embObjFTsensor`](https://github.com/robotology/icub-main/tree/v1.22.0/src/libraries/icubmod/embObjFTsensor) that read the FT indirectly via a ethernet connection to an EMS board, that then reads the value from the CAN bus (this is what is currently used on the latest version of the iCub robot.

**Reply by traversaro** [reply]
> But doesn't the repo also contains the ft sensor here. It is not fully clear to which kind of sensor the name 'forcetorque-yarp-devices/ftSensor' is referring."

That folder does not contain any YARP device driver (in the sense given in http://www.yarp.it/latest/dev-how-to.html, i.e. a C++ class that inherits from `yarp::dev::DeviceDriver`). It just contains two files from two experiments with different FT sensors.

**Comment by 2103simon** [comment]
And is there by any chance the way to use the same `canBusAnalogSensor`  script to read from the 12 pin connector of the nano ft sensor?

**Reply by traversaro** [reply]
The `canBusAnalogSensor` implements the iCub-specific CAN protocol that is implemented in the IIT's boards STRAIN and STRAIN2. I am not sure what signal are exposed in the 12 pin connector of the nano FT sensor, but I doubt that, even if it exposes a CAN device that implements a CAN protocol, that the ATI devices implements a iCub-specific CAN protocol.

**Reply by 2103simon** [reply]
I see... Do you have another suggestions on how to retrieve applied forces when not using the ATI nano? Or are you aware of a way to use the ATI nano?

**Reply by simeonedussoni** [reply]
We can have a look at the signals coming from the ATI sensor. It is possible theat they are some kindo f analog signals that can be handled by the STRAI2 electronics and thus streamed through its FW in the yarp flow. This option would require at least some calibration of the device to convert the analog values to F/T ones.

**Reply by 2103simon** [reply]
Hi @simeonedussoni, thanks for hopping in! That sounds even like an option! I am happy to help with what ever is needed. First I (we) can try to find out the nature of analog values and I can also help with designing the CAD model for the calibration procedure. Let me know if that is a possibility.

**Comment by pattacini** [comment]
Hi @2103simon, could you please stick to a thread while interacting with @traversaro instead of creating multiple answers? This way, we could keep the conversation neater. 

You can use the field <kbd>Write a reply</kbd> to this end.

Thanks!

**Reply by 2103simon** [reply]
OK, for sure! Sorry for inconvenience!

**Comment by 2103simon** [comment]
@simeonedussoni and me discussed today to create a setup to read out the analog values from the 12 pin connector the ATI nano ft sensor delivers and use electronics we have at place to make sense of it and on that track implement the nano sensor in my robotic setup.
@simeonedussoni will add needed equipment here and to me in person a short introduction and explanation on how to use it.

**Comment by lornat75** [comment]
hi @2103simon we used to read the ATI with a nidaq ADC, using a simple c library/API provided by NI, a few lines of code would be sufficient.

**Reply by 2103simon** [reply]
This works well in windows as far as I know (and I tried it using labView), but the omega robot I have to control only works on Linux. I also found the ni drivers for Linux, but here I can't read the sensor. I will investigate more. But maybe it could still be useful to have a fully YARP based read-out of the ATI ft sensor.
The driver for Linux i found here: https://www.ni.com/it-it/support/downloads/drivers/download.ni-daqmx-base.html#326057
